File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
       unordered_map<Node*, Node*> temp;
       Node* curr=head;
       while(curr){
            temp[curr] = new Node(curr->val);
            curr = curr->next;
       }
        curr = head;
       while(curr){
            Node* copy = temp[curr];
            copy->next = temp[curr->next];
            copy->random = temp[curr->random];
            curr = curr->next;
       }
       return temp[head];
    }
};

File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return TreeNode(val)
        curr = root
        while curr:
            temp = curr
            if val > curr.val:
                curr = curr.right
                if not curr:
                    temp.right = TreeNode(val)
            else:
                curr = curr.left
                if not curr:
                    temp.left = TreeNode(val)
        return root
            

